### PR TITLE
Apps to compute DORA metrics

### PR DESCRIPTION
**Pull request by snabar**
This set of apps compute the DORA metrics for a CI-CD pipeline for a customizable time window. The metrics include: 

* Mean Lead Time
* Mean Cycle Time
* Mean Time to Recovery
* Mean Time Between Failures
* Change Failure Rate

In addition to the metrics at the current point in time, the apps also compute the relative change from a previous point in time specified in the arguments.

### CODE REVIEW

File: apps/dora-metrics/recovery-time/app.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    deployments_current = list(
+        filter(lambda d: datetime.fromtimestamp(int(d["endedAt"])/1000) in current_window, deployments))
+
+    mrt_previous = mean_recovery_time(deployments_previous)
+    mrt_current = mean_recovery_time(deployments_current)
+
+    return {
+        "Mean Time to Recovery": mrt_current,
+        "Mean Recovery Time Change": relative_change(mrt_current, mrt_previous),
+    }
```
Comment (mtth) [inline apps/dora-metrics/recovery-time/app.py:74]: In the future, it could be nice to return the sample size too.

File: apps/dora-metrics/recovery-time/app.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import itertools
```
Comment (mtth) [inline apps/dora-metrics/recovery-time/app.py:1]: There's significant overlap between the apps here. What do you think about consolidating into a single dora app?

**Review by mtth — COMMENTED** [review]
Nice - it will be great to have these metrics readily available. Mostly code clarity comments.

File: apps/dora-metrics/cycle-time/app.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+from datetime import datetime
+from faros.utils import relative_change, get_comparison_windows
+from faros.client import FarosClient
+
+
+# TODO: Group by environment in addition to app name
+def mean_cycle_time(successful_deployments):
+    total_cycle_time = 0
+    num_apps = 0
+    for key, app in itertools.groupby(successful_deployments, lambda d: d["application"]["name"]):
```
[the remark this answers is not in the document]
Comment (snabar) [reply apps/dora-metrics/cycle-time/app.py:12]: Oh wow ... I didn't realize that. Does it make more sense to just keep track of the groups in a hash map then and avoid sorting a second time (the deployments are already sorted once by "ended at" timestamp).

File: apps/dora-metrics/recovery-time/app.py (hunk 2 above)
Comment (snabar) [reply apps/dora-metrics/recovery-time/app.py:1]: Following up on this over slack.

File: apps/dora-metrics/cycle-time/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtth) [reply apps/dora-metrics/cycle-time/app.py]: That works. Let's add a comment in the functions that make use of this assumption to avoid issues later on in that case. Something like:

```py
def mean_cycle_time(sorted_deploys):
  """Returns the average time between deployments. The input deployments must
  be grouped by app, and within each group sorted chronologically by end-time.
  """
  # ...
```

Otherwise sorting is cheap enough that it wouldn't be a significant issue to do it inside the function.

File: apps/dora-metrics/cycle-time/faros.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+appId: cycle-time
+displayName: Cycle Time
+description: "Computes the mean cycle time for a CI/CD pipeline. 
+The cycle time is aggregated over a period of <parameter window_days> days, and compared to the cycle time from <parameter compare_to_days> days ago.  
+For more information, https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance"
```
Comment (mtth) [inline apps/dora-metrics/cycle-time/faros.yaml:5]: YAML supports a different way of handling multi-line strings:

```yaml
description: >-
  Computes the mean cycle time for a CI/CD pipeline. The cycle time is
  aggregated over a period of <parameter window_days> days, and compared
  to the cycle time from <parameter compare_to_days> days ago. For more
  information, https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance
  
```

More info: https://yaml-multiline.info/

File: apps/dora-metrics/failure-rate/app.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    failed_deploys_previous = [d for d in deploys
+                               if d["status"] == "Failed" and datetime.fromtimestamp(int(d["endedAt"])/1000) in previous_window]
+    failed_deploys_current = [d for d in deploys
+                              if d["status"] == "Failed" and datetime.fromtimestamp(int(d["endedAt"])/1000) in current_window]
+
+    success_deploys_previous = [d for d in deploys
+                                if d["status"] == "Success" and datetime.fromtimestamp(int(d["endedAt"])/1000) in previous_window]
+    success_deploys_current = [d for d in deploys
+                               if d["status"] == "Success" and datetime.fromtimestamp(int(d["endedAt"])/1000) in current_window]
```
Comment (mtth) [inline apps/dora-metrics/failure-rate/app.py:61]: This would be good to extract to a function to avoid repetition here.

```py
def filter_deploys(deploys, status, window):
  # ...
```

File: apps/dora-metrics/lead-time/app.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        total_lead_time += mean_app_lead_time
+        num_apps += 1
+    return (total_lead_time/num_apps) * 1000 if num_apps else None
+
+
+# TODO: Loop through the builds as well?
+def filter_and_flatten(commits_with_builds):
+    deployed_commits = []
+    for c in commits_with_builds:
+        if (len(c["builds"]["nodes"][0]["deployments"]["nodes"]) > 0):
```
Comment (mtth) [inline apps/dora-metrics/lead-time/app.py:35]: Let's move the builds filtering (line 82) into this function. It makes this function easier to maintain by reducing assumptions on its inputs. As a side-benefit, the output of this filtering isn't used anywhere else so we will be able to keep it local here. It will also give this function a better defined purpose/name:

```py
def extract_deployed_commits(commits):
  # ...
```

File: apps/dora-metrics/recovery-time/app.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        num_app_recoveries = 0
+        for i in range(len(deploys)):
+            if deploys[i]["status"] == "Failed":
+                failure_time = int(deploys[i]["endedAt"])
+                for j in range(i+1, len(deploys)):
+                    if deploys[j]["status"] == "Success":
+                        success_time = int(deploys[j]["endedAt"])
+                        app_recovery_time += (success_time - failure_time)
+                        num_app_recoveries += 1
+                        i = j
```
Comment (mtth) [inline apps/dora-metrics/recovery-time/app.py:23]: Did you mean to use a `while` loop? Right now the assignment to `i` is ignored (and we don't skip over successful deployments).

File: apps/dora-metrics/cycle-time/faros.yaml (hunk 4 above)
Comment (snabar) [reply apps/dora-metrics/cycle-time/faros.yaml:5]: I did look at the guide above, and tried all four options ... the double quotes is what worked ... ">" gave me errors when I tried to build. Should I have done ">-" instead?  Is there a problem with using double quotes?

Comment (mtth) [reply apps/dora-metrics/cycle-time/faros.yaml:5]: What error are you getting? Could be a bug in how we parse YAML.

File: apps/dora-metrics/recovery-time/app.py (hunk 7 above)
Comment (snabar) [reply apps/dora-metrics/recovery-time/app.py:23]: good catch!

File: apps/dora-metrics/cycle-time/faros.yaml (hunk 4 above)
Comment (snabar) [reply apps/dora-metrics/cycle-time/faros.yaml:5]: ">-" does not work either. The error message is:
```
YAMLException: can not read a block mapping entry; a multiline key may not be an implicit key at line 7, column 8:
    handler: app.lambda_handler
```

In any case, what is the benefit of using this over " if both are supported?

Comment (mtth) [reply apps/dora-metrics/cycle-time/faros.yaml:5]: Are you indenting the lines of the description?

Trying locally:

```yml
# Works:
description: >-
  Computes change failure rate and mean time between failures in a CI/CD
  pipeline. The metrics are aggregated over a period of <parameter window_days>
  days, and compared to the metrics from <parameter compare_to_days> days ago.
  For more information, https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance

# Throws the error you link:
description: >-
Computes change failure rate and mean time between failures in a CI/CD
pipeline. The metrics are aggregated over a period of <parameter window_days>
days, and compared to the metrics from <parameter compare_to_days> days ago.
For more information, https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance
```


> In any case, what is the benefit of using this over " if both are supported?

Native multiline is easier to read and compatible with indent-based folds (used in many editors to navigate code - https://en.wikipedia.org/wiki/Code_folding).

File: apps/dora-metrics/cycle-time/test_app.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+from unittest.mock import patch
+
+
+# TODO: Add more tests for edge cases
+@patch.object(FarosClient, "graphql_execute")
+class CycleTimeTest(unittest.TestCase):
+    def setUp(self):
+        self.event = {"farosToken": "farosToken", "params": {
+            "window_days": 7, "compare_to_days": 7}}
+        self.date = datetime.now()
```
[the remark this answers is not in the document]
Comment (snabar) [reply apps/dora-metrics/cycle-time/test_app.py:16]: hmm ... the apps all compute and compare the metric value *now* to the metric value compare_to_days ago. So the tests unfortunately have to specify "now". i could add an extra parameter, which is the "end" time to be used instead of now, if that will be a common use case.

[the remark this answers is not in the document]
Comment (mtth) [reply apps/dora-metrics/cycle-time/test_app.py:16]: A parameter works. You can also patch it similar to `FarosClient.graphql_execute` above to avoid extending the public API. A few approaches here: https://stackoverflow.com/q/43799206/1062617.

[the remark this answers is not in the document]
Comment (snabar) [reply apps/dora-metrics/cycle-time/test_app.py]: Ah ... I went ahead and added the parameter everywhere, but I think I prefer the patching.

[the remark this answers is not in the document]
Comment (snabar) [reply apps/dora-metrics/cycle-time/test_app.py]: So I finally got the patching to work, but I think I'm gonna leave it in as a parameter since it offers more flexibility to the user.

File: apps/dora-metrics/recovery-time/app.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    Method expects deployments sorted by the application name and the end-time of the deployment
+    """
+    total_recovery_time = 0
+    num_apps = 0
+    for key, app in itertools.groupby(deployments, lambda d: d["application"]["name"]):
+        deploys = list(app)
+        app_recovery_time = 0
+        num_app_recoveries = 0
+        i = 0
+        while i in range(len(deploys)):
```
Comment (mtth) [inline apps/dora-metrics/recovery-time/app.py:19]: `while i < len(deploys):`

File: apps/dora-metrics/cycle-time/app.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+def mean_cycle_time(successful_deploys):
+    """
+    Method expects successful deployments sorted by the application name and the end-time of the deployment
+    """
+    span = 0
+    successive_deploys = 0
+    for key, app in itertools.groupby(successful_deploys, lambda d: d["application"]["name"]):
+        deploys = list(app)
+        if len(deploys) > 1:
+            span += (int(deploys[-1]["endedAt"]) - int(deploys[0]["endedAt"]))
```
Comment (mtth) [inline apps/dora-metrics/cycle-time/app.py:17]: You can omit the first parens here. Same just below and in other apps.

File: apps/dora-metrics/cycle-time/test_app.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                            }
+                        },
+                    ]
+                }
+            }
+        }
+        mock_execute.return_value = query_data
+        results = app.lambda_handler(self.event, None)
+        self.assertEqual(
+            results, {"Mean Cycle Time": 129600000, "Mean Cycle Time Change": -0.25})
```
Comment (mtth) [inline apps/dora-metrics/cycle-time/test_app.py:86]: `reference_time`? Same in other tests.

File: apps/dora-metrics/cycle-time/faros.yaml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+appId: cycle-time
+displayName: Cycle Time
+description: >
+  Computes the mean cycle time for a CI/CD pipeline. 
+  The cycle time is computed for a time window ending in <parameter reference_time_secs> which is a unix timestamp in seconds.
+  The default value for reference_time_secs is the current time. 
+  The metric is aggregated over a period of <parameter window_days> days, and compared to the cycle time from <parameter compare_to_days> days ago.
```
Comment (mtth) [inline apps/dora-metrics/cycle-time/faros.yaml:7]: Mention that it is weighted by deployment count?

File: apps/dora-metrics/lead-time/app.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        app_lead_time = 0
+        for c in commits:
+            app_lead_time += (c.deployment_ended_at -
+                              c.commit_created_at).total_seconds()
+        total_lead_time += app_lead_time
+        total_commits += len(commits)
+    return (total_lead_time/total_commits) * 1000 if total_commits else None
+
+
+# TODO: Loop through the builds as well?
```
Comment (mtth) [inline apps/dora-metrics/lead-time/app.py:33]: What is the reasoning for only using the first build?

File: apps/dora-metrics/lead-time/app.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+def extract_deployed_commits(commits):
+    deployed_commits = []
+    for c in commits:
+        if len(c["builds"]["nodes"]) > 0 and len(c["builds"]["nodes"][0]["deployments"]["nodes"]) > 0:
```
Comment (mtth) [inline apps/dora-metrics/lead-time/app.py:37]: You can omit the second condition, the loop below will do the right thing.

File: scripts/run_all_unittests.sh [hunk 15]
```diff
@@ -9,7 +9,15 @@ SCRIPT_DIR="$( cd -P "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
...
 APPS_BASE_DIR="${ROOT_DIR}/apps"
 
+for APP_DIR in ${APPS_BASE_DIR}/dora-metrics/* ; do
+		if [ -d "${APP_DIR}" ] ; then
+	    echo "Running tests in app directory ${APP_DIR}"
+  	  python3 -m unittest discover -s ${APP_DIR}
+		fi
+done
+
 for APP_DIR in ${APPS_BASE_DIR}/* ; do
```
Comment (mtth) [inline scripts/run_all_unittests.sh:19]: Is there a way to consolidate these two loops to find all tests, even nested (maybe with a **/*` glob)? It would help going forward, running tests in future directories. @chalenge 

Also, indentation looks off here. Mix of tabs and spaces?

**Review by mtth — APPROVED** [review]
Nice work!

File: scripts/run_all_unittests.sh (hunk 15 above)
Comment (snabar) [reply scripts/run_all_unittests.sh:19]: i did try that, but removed it cuz it ended up looking in a bunch of spurious folders like __pycache__ and node_modules folders ...

File: apps/dora-metrics/lead-time/app.py (hunk 14 above)
Comment (snabar) [reply apps/dora-metrics/lead-time/app.py:37]: ah ... good catch ... i had it left over from before i was actually looping over the deployments and just taking the first one :)

File: apps/dora-metrics/lead-time/app.py (hunk 13 above)
Comment (snabar) [reply apps/dora-metrics/lead-time/app.py:33]: Looping through now.

File: apps/dora-metrics/cycle-time/test_app.py (hunk 11 above)
Comment (snabar) [reply apps/dora-metrics/cycle-time/test_app.py:86]: I'm not following ... the mean cycle time is not the reference time.

Comment (mtth) [reply apps/dora-metrics/cycle-time/test_app.py:86]: You're right. Somehow I was expecting the constant to appear in the test.
